onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu

```diff
@@ -72,70 +80,293 @@ __global__ void SkipLayerNormKernel(
-    cudaStream_t stream, const int ld, const int n, const T* input, const T* skip,
-    const T* beta, const T* gamma, const T* bias, const T epsilon, T* output) {
+template <unsigned TPB, int ILP>
+__global__ void SkipLayerNormKernelSmallVec(
+    const int ld, const half* input, const half* skip, const half* beta, const half* gamma,
+    const half* bias, const half epsilon, half* output, bool hasBias) {
+  const half rld = half(1.f) / half(ld);
+  const int idx = blockIdx.x * ld + threadIdx.x * ILP;  // grid_size = n / ld
+
+  using VecT = aligned_vector<half, ILP>;
+  using BlockReduce = cub::BlockReduce<cub::KeyValuePair<half, half>, TPB>;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+  __shared__ half mu;      // mean
+  __shared__ half rsigma;  // 1 / std.dev.
+  KeyValuePairSum pair_sum;
+
+  using VecT = aligned_vector<half, ILP>;
+  half input_v[ILP], skip_v[ILP], beta_v[ILP], gamma_v[ILP], bias_v[ILP], output_v[ILP];
+
+  VecT* input_val = reinterpret_cast<VecT*>(&input_v);
+  *input_val = *reinterpret_cast<const VecT*>(&input[idx]);
+
+  VecT* skip_val = reinterpret_cast<VecT*>(&skip_v);
+  *skip_val = *reinterpret_cast<const VecT*>(&skip[idx]);
+
+  if (beta != nullptr) {
+    VecT* beta_val = reinterpret_cast<VecT*>(&beta_v);
+    *beta_val = *reinterpret_cast<const VecT*>(&beta[threadIdx.x * ILP]);
+  }
+
+  VecT* gamma_val = reinterpret_cast<VecT*>(&gamma_v);
+  *gamma_val = *reinterpret_cast<const VecT*>(&gamma[threadIdx.x * ILP]);
+
+  if (hasBias) {
+    VecT* bias_val = reinterpret_cast<VecT*>(&bias_v);
+    *bias_val = *reinterpret_cast<const VecT*>(&bias[threadIdx.x * ILP]);
+  }
+
+  VecT* output_val = reinterpret_cast<VecT*>(&output_v);
+
+  cub::KeyValuePair<half, half> thread_data;
+  if (ILP * threadIdx.x < ld) {
+    half rldval_sum = half(0.f);
+    half rldvalsq_sum = half(0.f);
+    #pragma unroll
+    for (int i = 0; i < ILP; i++) {
+      input_v[i] += skip_v[i];
+      if (hasBias) {
+        input_v[i] += bias_v[i];
+      }
+      const half rldval = rld * input_v[i];
+      rldval_sum += rldval;
+      rldvalsq_sum += rldval * input_v[i];
+    }
+    thread_data = cub::KeyValuePair<half, half>(rldval_sum, rldvalsq_sum);
+  } else {
+    thread_data = cub::KeyValuePair<half, half>(half(0.f), half(0.f));
+  }
+  const cub::KeyValuePair<half, half> sumKV = BlockReduce(temp_storage).Reduce(thread_data, pair_sum);
+  if (threadIdx.x == 0) {
+    mu = sumKV.key;
+    rsigma = Rsqrt(sumKV.value - mu * mu + epsilon);
+  }
+  __syncthreads();
+  if (ILP * threadIdx.x < ld) {
+    #pragma unroll
+    for (int i = 0; i < ILP; i++) {
+      output_v[i] = (beta != nullptr) ? gamma_v[i] * (input_v[i] - mu) * rsigma + beta_v[i] :
+                                        gamma_v[i] * (input_v[i] - mu) * rsigma;
+    }
+    *(reinterpret_cast<VecT*>(&output[idx])) = *reinterpret_cast<VecT*>(&output_v[0]);
+  }
+}
+
+template <unsigned TPB, int ILP>
+__global__ void SkipLayerNormKernelVec(
```

Comment:
Are the code of SkipLayerNormKernelSmallVec and SkipLayerNormKernelVec pretty much the same now? Which parts of the code are different?